### PR TITLE
feat: supported tools/functions in claude 2.1

### LINKED ISSUE #38: Support tools/functions in Claude 2.1
* For Claude 2.1 see the [demo](https://docs.anthropic.com/claude/docs/guide-to-anthropics-prompt-engineering-resources#function-calling--tool-use-repo) with the recommended prompts.
* Llama apparently works well with prompts from Langchain (see `langchain.agents` and `chat-conversational-react-description` agent with `ConversationBufferWindowMemory` memory).

### PR DESCRIPTION
**Pull request by adubovik**
Resolves #38

### CODE REVIEW

File: aidial_adapter_bedrock/utils/xml.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+def tag_nl(name: str, arg: str) -> str:
+    ...
+
+
+def tag_nl(name: str, arg: Arg) -> str | None:
+    content = _arg_to_str(arg)
+    if content is None:
+        return None
+    content = "\n" if content == "" else f"\n{content}\n"
+    return f"<{name}>{content}</{name}>" if content is not None else None
```
Comment (Oleksii-Klimov) [inline aidial_adapter_bedrock/utils/xml.py]: Redundant _if_

File: aidial_adapter_bedrock/llm/tools/claude_emulator.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    mode: Optional[ToolsMode] = tool_config and tool_config.mode
+    match msg:
+        case HumanToolResultMessage(id=id, content=content):
+            assert (
+                mode is None or mode == ToolsMode.TOOLS
+            ), f"Received tool result in '{mode.value}' mode"
+            name = id_to_name.get(id)
+            if name is None:
+                log.warning(
+                    f"Unable to find tool name for id '{id}', assuming '_unknown' name"
```
Comment (Oleksii-Klimov) [inline aidial_adapter_bedrock/llm/tools/claude_emulator.py]: Extra _

File: aidial_adapter_bedrock/llm/message.py [hunk 3]
```diff
@@ -1,48 +1,92 @@
...
+            function_call=function_call,
+            tool_calls=tool_calls,
+        ):
+            return _parse_assistant_message(content, function_call, tool_calls)
+        case Message(
+            role=Role.FUNCTION, name=name, content=content
+        ) if content is not None and name is not None:
+            return HumanFunctionResultMessage(name=name, content=content)
+        case Message(
+            role=Role.TOOL, tool_call_id=id, content=content
```
Comment (Oleksii-Klimov) [inline aidial_adapter_bedrock/llm/message.py:88]: Will it work if name is set? Here in the cookbook examples the name is set to function name: https://cookbook.openai.com/examples/how_to_call_functions_with_chat_models

**Review by Oleksii-Klimov — DISMISSED** [review]

File: aidial_adapter_bedrock/utils/xml.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_bedrock/utils/xml.py]: Fixed

File: aidial_adapter_bedrock/llm/tools/claude_emulator.py (hunk 2 above)
Comment (adubovik) [reply aidial_adapter_bedrock/llm/tools/claude_emulator.py]: Fixed

File: aidial_adapter_bedrock/llm/message.py (hunk 3 above)
Comment (adubovik) [reply aidial_adapter_bedrock/llm/message.py:88]: The notebook seems to be outdated. The OpenAI API doesn't declare `name` field in the Tool Message: https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages

`name` is redundant - you could figure out the function name by looking at the `tool_call_id` field.

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Review by Oleksii-Klimov — APPROVED** [review]

**Pull request merged**